File: backend/src/main/java/com/tripmate/service/CompanionService.java
package com.tripmate.service;

import com.tripmate.dto.CompanionApplicationDto;
import com.tripmate.dto.CompanionDto;
import com.tripmate.entity.*;
import com.tripmate.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CompanionService {

    private final CompanionRepository companionRepository;
    private final CompanionApplicationRepository applicationRepository;
    private final TripRepository tripRepository;
    private final ChatRoomRepository chatRoomRepository;

    @Transactional(readOnly = true)
    public Page<CompanionDto> getCompanions(Pageable pageable) {
        return companionRepository.findAllByOrderByCreatedAtDesc(pageable)
                .map(CompanionDto::from);
    }

    @Transactional(readOnly = true)
    public CompanionDto getCompanion(Long id) {
        Companion companion = companionRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Companion not found"));
        return CompanionDto.from(companion);
    }

    @Transactional
    public CompanionDto createCompanion(User user, CompanionDto.CreateRequest request) {
        Trip trip = tripRepository.findById(request.getTripId())
                .orElseThrow(() -> new IllegalArgumentException("Trip not found"));

        if (!trip.getUser().getId().equals(user.getId())) {
            throw new IllegalArgumentException("Not authorized");
        }

        Companion companion = Companion.builder()
                .trip(trip)
                .user(user)
                .title(request.getTitle())
                .content(request.getContent())
                .maxMembers(request.getMaxMembers())
                .build();

        companion = companionRepository.save(companion);

        ChatRoom chatRoom = ChatRoom.builder()
                .companion(companion)
                .build();
        chatRoomRepository.save(chatRoom);

        return CompanionDto.from(companion);
    }

    @Transactional
    public CompanionDto updateCompanion(Long id, User user, CompanionDto.UpdateRequest request) {
        Companion companion = companionRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Companion not found"));

        if (!companion.getUser().getId().equals(user.getId())) {
            throw new IllegalArgumentException("Not authorized");
        }

        if (request.getTitle() != null) companion.setTitle(request.getTitle());
        if (request.getContent() != null) companion.setContent(request.getContent());
        if (request.getMaxMembers() != null) companion.setMaxMembers(request.getMaxMembers());
        if (request.getStatus() != null) companion.setStatus(Companion.CompanionStatus.valueOf(request.getStatus()));

        return CompanionDto.from(companionRepository.save(companion));
    }

    @Transactional
    public void deleteCompanion(Long id, User user) {
        Companion companion = companionRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Companion not found"));

        if (!companion.getUser().getId().equals(user.getId())) {
            throw new IllegalArgumentException("Not authorized");
        }

        companionRepository.delete(companion);
    }

    @Transactional
    public CompanionApplicationDto apply(Long companionId, User user, String message) {
        Companion companion = companionRepository.findById(companionId)
                .orElseThrow(() -> new IllegalArgumentException("Companion not found"));

        if (companion.getUser().getId().equals(user.getId())) {
            throw new IllegalArgumentException("Cannot apply to your own companion");
        }

        if (companion.getStatus() != Companion.CompanionStatus.RECRUITING) {
            throw new IllegalArgumentException("Not recruiting");
        }

        if (applicationRepository.existsByCompanionAndUser(companion, user)) {
            throw new IllegalArgumentException("Already applied");
        }

        CompanionApplication application = CompanionApplication.builder()
                .companion(companion)
                .user(user)
                .message(message)
                .build();

        return CompanionApplicationDto.from(applicationRepository.save(application));
    }

    @Transactional(readOnly = true)
    public List<CompanionApplicationDto> getApplications(Long companionId, User user) {
        Companion companion = companionRepository.findById(companionId)
                .orElseThrow(() -> new IllegalArgumentException("Companion not found"));

        if (!companion.getUser().getId().equals(user.getId())) {
            throw new IllegalArgumentException("Not authorized");
        }

        return applicationRepository.findByCompanionOrderByCreatedAtDesc(companion)
                .stream()
                .map(CompanionApplicationDto::from)
                .collect(Collectors.toList());
    }

    @Transactional
    public void approveApplication(Long companionId, Long userId, User currentUser) {
        Companion companion = companionRepository.findById(companionId)
                .orElseThrow(() -> new IllegalArgumentException("Companion not found"));

        if (!companion.getUser().getId().equals(currentUser.getId())) {
            throw new IllegalArgumentException("Not authorized");
        }

        CompanionApplication application = applicationRepository
                .findByCompanionAndUser(companion, User.builder().id(userId).build())
                .orElseThrow(() -> new IllegalArgumentException("Application not found"));

        application.setStatus(CompanionApplication.ApplicationStatus.APPROVED);
        applicationRepository.save(application);

        companion.setCurrentMembers(companion.getCurrentMembers() + 1);
        if (companion.getCurrentMembers() >= companion.getMaxMembers()) {
            companion.setStatus(Companion.CompanionStatus.CLOSED);
        }
        companionRepository.save(companion);
    }

    @Transactional
    public void rejectApplication(Long companionId, Long userId, User currentUser) {
        Companion companion = companionRepository.findById(companionId)
                .orElseThrow(() -> new IllegalArgumentException("Companion not found"));

        if (!companion.getUser().getId().equals(currentUser.getId())) {
            throw new IllegalArgumentException("Not authorized");
        }

        CompanionApplication application = applicationRepository
                .findByCompanionAndUser(companion, User.builder().id(userId).build())
                .orElseThrow(() -> new IllegalArgumentException("Application not found"));

        application.setStatus(CompanionApplication.ApplicationStatus.REJECTED);
        applicationRepository.save(application);
    }
}

File: frontend/src/services/companionService.ts
import api from './api'
import type { Companion, CompanionApplication, ApiResponse, PageResponse } from '@/types'

interface CreateCompanionRequest {
  tripId: number
  title: string
  content: string
  maxMembers: number
}

interface UpdateCompanionRequest extends Partial<CreateCompanionRequest> {
  status?: string
}

interface CompanionFilters {
  destination?: string
  theme?: string
  startDate?: string
  endDate?: string
}

export const companionService = {
  getCompanions: async (page = 0, size = 10, filters?: CompanionFilters): Promise<PageResponse<Companion>> => {
    const response = await api.get<ApiResponse<PageResponse<Companion>>>('/companions', {
      params: { page, size, ...filters },
    })
    return response.data.data
  },

  getCompanion: async (id: number): Promise<Companion> => {
    const response = await api.get<ApiResponse<Companion>>(`/companions/${id}`)
    return response.data.data
  },

  createCompanion: async (data: CreateCompanionRequest): Promise<Companion> => {
    const response = await api.post<ApiResponse<Companion>>('/companions', data)
    return response.data.data
  },

  updateCompanion: async (id: number, data: UpdateCompanionRequest): Promise<Companion> => {
    const response = await api.put<ApiResponse<Companion>>(`/companions/${id}`, data)
    return response.data.data
  },

  deleteCompanion: async (id: number): Promise<void> => {
    await api.delete(`/companions/${id}`)
  },

  apply: async (companionId: number, message: string): Promise<CompanionApplication> => {
    const response = await api.post<ApiResponse<CompanionApplication>>(`/companions/${companionId}/apply`, {
      message,
    })
    return response.data.data
  },

  getApplications: async (companionId: number): Promise<CompanionApplication[]> => {
    const response = await api.get<ApiResponse<CompanionApplication[]>>(`/companions/${companionId}/applications`)
    return response.data.data
  },

  approveApplication: async (companionId: number, userId: number): Promise<void> => {
    await api.post(`/companions/${companionId}/approve/${userId}`)
  },

  rejectApplication: async (companionId: number, userId: number): Promise<void> => {
    await api.post(`/companions/${companionId}/reject/${userId}`)
  },
}
